Rework a number of SIL and IRGen witness-table abstractions

rjmccall opened the pull request:
Rework a number of SIL and IRGen witness-table abstractions to correctly handle generalized protocol requirements.

The major missing pieces here are that the conformance search algorithms in both the AST (type substitution) and IRGen (witness table reference emission) need to be rewritten to back-track requirement sources, and the AST needs to actually represent this stuff in NormalProtocolConformances instead of just doing ???.

The new generality isn't tested yet; I'm looking into that, but I wanted to get the abstractions in place first.

rjmccall committed:
Rework a number of SIL and IRGen witness-table abstractions
to correctly handle generalized protocol requirements.

The major missing pieces here are that the conformance search
algorithms in both the AST (type substitution) and IRGen
(witness table reference emission) need to be rewritten to
back-track requirement sources, and the AST needs to actually
represent this stuff in NormalProtocolConformances instead
of just doing ???.

The new generality isn't tested yet; I'm looking into that,
but I wanted to get the abstractions in place first.

rjmccall commented:
@swift-ci Please test and merge.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
+
+  // Search the protocol.
+  for (auto member : protocol->getMembers()) {
+    if (auto memberAssoc = dyn_cast<AssociatedTypeDecl>(member)) {
```
Should this also handle typealiases?

rjmccall replied:
They won't be used as keys in the TypeWitnesses map, so I don't think that's necessary.  This is really just a temporary workaround around the bad internal representation of NormalProtocolConformance, which will need to be improved to handle generalized requirements anyway — it's literally not even storing conformance references for associated paths of length > 1.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
+  // Fast path.
+  auto type = assocType->getCanonicalType();
+  if (isa<GenericTypeParamType>(type)) {
+    return getType();
```
Assert that its equal to the protocol Self type (proto->getSelfInterfaceType())

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
+         "associated type must be a type parameter");
+
+  // Fast path.
+  auto type = assocType->getCanonicalType();
```
Why are you canonicalizing it here?

slavapestov commented on lib/Serialization/SILFormat.h:
```diff
@@ -200,7 +200,7 @@ namespace sil_block {
   using WitnessAssocProtocolLayout = BCRecordLayout<
     SIL_WITNESS_ASSOC_PROTOCOL,
-    DeclIDField, // ID of AssociatedTypeDecl
+    TypeIDField, // ID of associated type
```
ModuleFormat version needs to be incremnted

rjmccall replied:
Of course, will fix.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
 }
 
+#ifndef NDEBUG
+static bool isSelfParameter(CanType type) {
```
This will cause crashes on invalid code. Instead, always compare against proto->getSelfInterfaceType().

rjmccall replied:
Will do (and in other places where reasonable).

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
+NormalProtocolConformance::getAssociatedConformance(Type assocType,
+                                                    ProtocolDecl *protocol,
+                                                LazyResolver *resolver) const {
+  assert(assocType->isTypeParameter() &&
```
This feels like it's duplicating work that's already in SubstitutionMap and GenericSignatureBuilder

rjmccall replied:
It absolutely may be.  If you can improve the implementation, please do.  I was trying to implement an operation that something like SubstitutionMap could use as a primitive.

That said, please do try to design APIs so that people don't need to learn all of the innards of a billion other types in order to get work done.  There's no reason that ProtocolConformance shouldn't have an API to translate an arbitrary type or conformance requirement from the protocol context into the conforming context.

DougGregor commented:
https://github.com/apple/swift/pull/7853 adds the actual conformances that line up with the requirement signature to `NormalProtocolConformance`, so we don't have to hunt for them at emission time

DougGregor commented on include/swift/SIL/SILWitnessVisitor.h:
```diff
@@ -49,27 +49,76 @@ template <class T> class SILWitnessVisitor : public ASTVisitor<T> {
+        addAssociatedTypes();
+
+        // Otherwise, add an associated requirement.
+        asDerived().addAssociatedConformance(type, protocol);
```
With https://github.com/apple/swift/pull/7853, we can provide the actual conformance here without having to go look for it. Does that obviate the need for `NormalProtocolConformance::getAssociatedConformance()`?

rjmccall replied:
Is this comment supposed to be somewhere else?  SILWitnessVisitor is an abstract operation over protocols, not protocol conformances.

rjmccall commented:
@swift-ci Please smoke test and merge.

rjmccall commented:
@swift-ci Please smoke test.

rjmccall commented:
I guess swift-ci is unhappy right now.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

rjmccall commented:
Minimal test case for Linux failure:

red1.swift: protocol P {}
red2.swift: func foo() -> P? { return nil }
swiftc -frontend -c red1.swift -primary-file red2.swift

rjmccall commented:
@DougGregor, I appreciate you trying to merge this, but it looks like the linux builder found a legit bug, albeit one I'm not sure is in my patch. :)

rjmccall committed:
Rename ValidatedTypes/validateTypeForLayout to make the
phase ordering self-documenting.  NFC.

rjmccall committed:
Fix a bug introduced by the recursive-protocol validation
special case by ensuring that protocols are fully validated
before Sema completes.

The test case doesn't currently fail without this patch, but
there are pending changes which will cause SILGen and IRGen
to depend on the requirement signature existing.

rjmccall commented:
Alright, we've got a fix.  Let's try this again.

@swift-ci Please smoke test and merge.

rjmccall commented:
@swift-ci please smoke test and merge
